src/xrpld/app/misc/NetworkOPs.cpp

```diff
@@ -1396,16 +1493,28 @@ NetworkOPsImp::apply(std::unique_lock<std::mutex>& batchLock)
                     << "Transaction is now included in open ledger";
                 e.transaction->setStatus(INCLUDED);
 
+                // Pop as many "reasonable" transactions for this account as
+                // possible. "Reasonable" means they have sequential sequence
+                // numbers, or use tickets.
                 auto const& txCur = e.transaction->getSTransaction();
-                auto const txNext = m_ledgerMaster.popAcctTransaction(txCur);
-                if (txNext)
+
+                std::size_t count = 0;
+                for (auto txNext = m_ledgerMaster.popAcctTransaction(txCur);
+                     txNext && count < maxPoppedTransactions;
```

Comment:
If we only pop a maximum of ten transactions, does this mean that over time, we will have a slow memory leak here of transactions that will never be used but will stay in the CanonicalTXSet indefinitely?

Reply:
No, because the `mHeldTransactions` list that this function pulls from is also fully cleared out and retried all at once in `LedgerMaster::applyHeldTransactions()`. `applyHeldTransactions` is called during consensus when a new closed ledger is created from the open ledger to start the process.

There are restrictions on how many times a transaction can be put into `mHeldTransactions` so that it can't cycle in and out indefinitely. Those are described at one of the call sites of `addHeldTransaction()`